### PR TITLE
[netcore] Port classes to netcore

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

**Comment by vargaz** [comment]
With these changes, execution goes until RuntimeType.GetConstructors_internal ().

**Comment by vargaz** [comment]
https://github.com/mono/mono/pull/12549

File: mono/native/Makefile.am [hunk 1]
```diff
@@ -34,7 +34,9 @@ common_sources = \
 	../../external/corefx/src/Native/Unix/System.Native/pal_uid.c \
 	../../external/corefx/src/Native/Unix/System.Native/pal_uid.h \
 	../../external/corefx/src/Native/Unix/System.Native/pal_time.c \
-	../../external/corefx/src/Native/Unix/System.Native/pal_time.h
+	../../external/corefx/src/Native/Unix/System.Native/pal_time.h \
```
Comment (marek-safar) [inline mono/native/Makefile.am]: We should not load existing Mono.Native at all, why is it needed?

File: mono/native/pal_config.h [hunk 2]
```diff
@@ -11,3 +11,22 @@
 /* arc4random_buf() is not available even when configure seems to find it */
 #undef HAVE_ARC4RANDOM_BUF
 #endif
+
```
Comment (marek-safar) [inline mono/native/pal_config.h]: same as above

**Comment by vargaz** [comment]
This now goes until:
System.TypeInitializationException: The type initializer for 'System.Globalization.CompareInfo' threw an exception. ---> System.Globalization.Native: Dll was not found.

File: mcs/class/System.Private.CoreLib/System.Collections.Generic/EqualityComparer.cs [hunk 3]
```diff
@@ -1,17 +1,58 @@
...
+
 namespace System.Collections.Generic
 {
 	partial class EqualityComparer<T>
 	{
-		public static EqualityComparer<T> Default {
-			get {
-				throw new NotImplementedException ();
-			}
+        static volatile EqualityComparer<T> defaultComparer;
```
Comment (filipnavara) [inline mcs/class/System.Private.CoreLib/System.Collections.Generic/EqualityComparer.cs]: Nit: The indentation is off in the whole file. Makes it hard to read on GitHub web.

File: mono/native/Makefile.am (hunk 1 above)
Comment (lambdageek) [reply mono/native/Makefile.am]: @vargaz @marek-safar I didn't update the native library loader to use the paths from the coreclr_initialize property list yet - so right now Mono is still trying to resolve .dylib loads against $(prefix), not `NATIVE_DLL_SEARCH_DIRECTORIES` or `APP_NI_PATHS`  (see https://github.com/mono/mono/blob/55d29aa7d4038846b12ce1b9350ccab738b48a2f/mono/mini/main-core.c#L120-L123)

**Comment by vargaz** [comment]
Remove the native changes, some of them seem broken on some platforms.

**Comment by vargaz** [comment]
@monojenkins commit apidiff

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

File: mcs/class/System.Private.CoreLib/System.Globalization/GlobalizationMode.cs [hunk 4]
```diff
@@ -6,6 +6,19 @@ namespace System.Globalization
...
+			var val = Environment.GetEnvironmentVariable ("DOTNET_SYSTEM_GLOBALIZATION_INVARIANT");
+			if (val != null)
+				return Boolean.IsTrueStringIgnoreCase (val) || val.Equals ("1");
+			int res = Interop.Globalization.LoadICU ();
+			if (res == 0) {
+				string message = "Couldn't find a valid ICU package installed on the system. " +
+					"Set the configuration flag System.Globalization.Invariant to true if you want to run with no globalization support.";
+				Environment.FailFast (message);
+			}
+            return false;
```
Comment (filipnavara) [inline mcs/class/System.Private.CoreLib/System.Globalization/GlobalizationMode.cs]: Nit: Mixed TABs and spaces.

File: mcs/class/System.Private.CoreLib/System/Environment.cs [hunk 5]
```diff
@@ -1,7 +1,14 @@
+using System;
 using System.Globalization;
+using System.Runtime.CompilerServices;
+using Mono;
 
 namespace System
 {
+    enum PlatformID
+    {
+    }
```
Comment (filipnavara) [inline mcs/class/System.Private.CoreLib/System/Environment.cs]: Nit: Wrong indentation?

File: mcs/class/corlib/System.Reflection/MonoMethod.cs [hunk 6]
```diff
@@ -44,6 +45,42 @@
 using System.Diagnostics;
 
 namespace System.Reflection {
+
+#if NETCORE
+	[Flags()]
+    internal enum PInvokeAttributes
```
Comment (filipnavara) [inline mcs/class/corlib/System.Reflection/MonoMethod.cs]: Nit: Broken indentation

File: mcs/class/corlib/System.Reflection/MonoMethod.cs [hunk 7]
```diff
@@ -719,6 +778,9 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
         {
             if (info == null)
                 throw new ArgumentNullException("info");
+#if NETCORE
+			throw new NotImplementedException ();
```
Comment (filipnavara) [inline mcs/class/corlib/System.Reflection/MonoMethod.cs]: Nit: Broken indentation

File: mcs/class/corlib/System.Reflection/MonoMethod.cs [hunk 8]
```diff
@@ -727,12 +789,17 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
                 SerializationToString(),
                 MemberTypes.Constructor,
                 null);
+#endif
         }
 
         internal string SerializationToString()
         {
+#if NETCORE
+			throw new NotImplementedException ();
```
Comment (filipnavara) [inline mcs/class/corlib/System.Reflection/MonoMethod.cs]: ...and here...

File: mcs/class/corlib/System.Reflection/MonoProperty.cs [hunk 9]
```diff
@@ -112,6 +116,9 @@ public override String ToString()
 
         private string FormatNameAndSig(bool serialization)
         {
+#if NETCORE
+			throw new NotImplementedException ();
```
Comment (filipnavara) [inline mcs/class/corlib/System.Reflection/MonoProperty.cs]: ...and here...

File: mcs/class/referencesource/mscorlib/system/rttype.cs [hunk 10]
```diff
@@ -3235,10 +3237,14 @@ public override InterfaceMapping GetInterfaceMap(Type ifaceType)
                     for (int j = 1; j < candidates.Count; j++)
                     {
                         MethodInfo methodInfo = candidates[j];
+#if NETCORE
+						throw new NotImplementedException ();
```
Comment (filipnavara) [inline mcs/class/referencesource/mscorlib/system/rttype.cs]: ...and here...

File: mcs/class/referencesource/mscorlib/system/rttype.cs [hunk 11]
```diff
@@ -4272,7 +4287,11 @@ public override Type MakeGenericType(Type[] instantiation)
                     for (int iCopy = 0; iCopy < instantiation.Length; iCopy++)
                         instantiationCopy[iCopy] = instantiation[iCopy];
                     instantiation = instantiationCopy;
+#if NETCORE
+					throw new NotImplementedException ();
```
Comment (filipnavara) [inline mcs/class/referencesource/mscorlib/system/rttype.cs]: ...and here...

File: mcs/class/referencesource/mscorlib/system/rttype.cs [hunk 12]
```diff
@@ -5395,7 +5421,11 @@ private void CreateInstanceCheckThis()
                             } else {
 #endif
                             // fast path??
+#if NETCORE
+								throw new NotImplementedException ();
```
Comment (filipnavara) [inline mcs/class/referencesource/mscorlib/system/rttype.cs]: ...and here...

**Comment by vargaz** [comment]
@monojenkins build failed

File: mcs/class/System.Private.CoreLib/System/Environment.cs [hunk 13]
```diff
@@ -77,5 +95,18 @@ public static void FailFast(string message, Exception exception)
 		public static void FailFast (string message, Exception exception, string errorMessage)
 		{
 		}
+
+		static internal PlatformID Platform {
+			[MethodImplAttribute (MethodImplOptions.InternalCall)]
+			get;
+		}
+
+#if (MONOTOUCH || MONODROID || XAMMAC)
```
[the remark this answers is not in the document]
Comment (lambdageek) [reply mcs/class/System.Private.CoreLib/System/Environment.cs]: It's just copypasted from corlib's System.Environment - so just optimistic futureproofing.

File: mcs/class/corlib/ReferenceSources/RuntimeType.cs [hunk 14]
```diff
@@ -544,14 +550,14 @@ public override InterfaceMapping GetInterfaceMap (Type ifaceType)
 
 			InterfaceMapping res;
 			if (!ifaceType.IsInterface)
-				throw new ArgumentException (Locale.GetText ("Argument must be an interface."), "ifaceType");
+				throw new ArgumentException ("Argument must be an interface.", "ifaceType");
```
Comment (lambdageek) [inline mcs/class/corlib/ReferenceSources/RuntimeType.cs]: I don't understand why it's ok to remove these calls to `Locale.GetText` - isn't this changing something for our corlib?

File: mcs/class/System.Private.CoreLib/System/RuntimeType.cs [hunk 15]
```diff
@@ -229,7 +229,15 @@ public override MethodInfo[] GetMethods(BindingFlags bindingAttr)
 			throw new NotImplementedException (); /* TODO */
 		}
 
-
+		internal static object CreateInstanceForAnotherGenericParameter (Type genericType, RuntimeType genericArgument)
```
Comment (lambdageek) [inline mcs/class/System.Private.CoreLib/System/RuntimeType.cs]: isn't this entire file dead code? I saw you removed `System/RuntimeType.cs` from `System.Private.CoreLib.csproj`

File: mcs/class/System.Private.CoreLib/System.Collections.Generic/EqualityComparer.cs [hunk 16]
```diff
@@ -1,17 +1,58 @@
+using System.Runtime.CompilerServices;
```
Comment (lambdageek) [inline mcs/class/System.Private.CoreLib/System.Collections.Generic/EqualityComparer.cs]: EqualityComparer.cs isn't added to `System.Private.CoreLib.csproj` - so this is dead code right now.

Comment (vargaz) [reply mcs/class/System.Private.CoreLib/System.Collections.Generic/EqualityComparer.cs]: It is there:
      <Compile Include="System.Collections.Generic\EqualityComparer.cs" />

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Pull request merged**